src/Database/Persist/URL.hs

```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Database.Persist.URL
+    ( fromDatabaseUrl
+    ) where
+
+import Control.Monad (MonadPlus, mzero)
+import Data.ByteString (ByteString, uncons)
+import Data.Monoid ((<>))
+import Database.Persist.Postgresql (PostgresConf(..))
+import URI.ByteString
+    ( Authority(..)
+    , Host(..)
+    , Port(..)
+    , URI(..)
+    , UserInfo(..)
+    , parseURI
+    , strictURIParserOptions
+    )
+
+import qualified Data.ByteString.Char8 as Char8
+
+-- | Build a @'PostgresConf'@ by parsing a database URL String
+fromDatabaseUrl :: MonadPlus m => Int -> ByteString -> m PostgresConf
```

Comment:
For those keeping track at home: The `MonadFail` instance for `Parser` didn't show up in Aeson until 0.11, which still isn't in Stackage.

This means that, for Yesod applications, there currently isn't any practical benefit to using `MonadFail` like this.